0.preprocess-sites/8.visualize-plate-layout.py

```diff
@@ -0,0 +1,392 @@
+        break
+    else:
+        build_seq.append([x_seq, y_seq])
+        if len(build_seq) == acquisition_side:
+            x_seq = x_seq + 1
+        if y_seq == acquisition_side:
+            y_seq = 0
+            build_seq = build_seq[::-1]
+            final_order.extend(build_seq)
+            build_seq = []
+        y_seq = y_seq + 1
+sites_list = [*range(1, (acquisition_side * acquisition_side) + 1)]
+
+# uses sites_list in case there are fewer analyzed sites than acquired sites
+loc_df = pd.DataFrame(final_order).rename(columns={0: "x_loc", 1: "y_loc"})
+loc_df["Site"] = sites_list
+
+# create total_cell_count
+cell_count_bysite_df = (
+    cell_count_df.groupby("site")["cell_count"]
+    .sum()
+    .reset_index()
+    .rename(columns={"cell_count": "total_cell_count"})
+)
+# add total_cell_count to cell_count_df
+cell_count_df = cell_count_df.merge(cell_count_bysite_df, on="site")
+# add in x, y coordinates for plotting
+cell_count_df = cell_count_df.merge(loc_df, left_on="Site", right_on="Site")
+
+# Plot total number of cells per well
+cell_count_totalcells_df = (
+    cell_count_df.groupby(["x_loc", "y_loc", "Well", "Site"])["total_cell_count"]
+    .mean()
+    .reset_index()
+)
+os.makedirs(output_figuresdir, exist_ok=True)
+
+by_well_gg = (
+    gg.ggplot(cell_count_totalcells_df, gg.aes(x="x_loc", y="y_loc"))
+    + gg.geom_point(gg.aes(fill="total_cell_count"), size=10)
+    + gg.geom_text(gg.aes(label="Site"), color="lightgrey")
+    + gg.facet_wrap("~Well")
+    + gg.coord_fixed()
+    + gg.theme_bw()
+    + gg.theme(
+        axis_text=gg.element_blank(),
+        axis_title=gg.element_blank(),
+        strip_background=gg.element_rect(colour="black", fill="#fdfff4"),
+    )
+    + gg.scale_fill_cmap(name="magma")
+)
+output_file = pathlib.Path(output_figuresdir, "plate_layout_cells_count_per_well.png")
+by_well_gg.save(output_file, dpi=300, verbose=False)
+
+# Plot cell category ratios per well
+ratio_df = pd.pivot_table(
+    cell_count_df,
+    values="cell_count",
+    index=["site", "Plate", "Well", "Site", "x_loc", "y_loc"],
+    columns=["Cell_Quality"],
+)
+ratio_df["Sum"] = ratio_df.sum(axis=1)
+ratio_df["Pass_Filter"] = ratio_df[cell_filter].sum(axis=1)
+fail_filter = [cat for cat in cell_category_order if cat not in cell_filter]
+ratio_df["Fail_Filter"] = ratio_df[fail_filter].sum(axis=1)
+fail_filter_noempty = fail_filter.remove("Empty")
+ratio_df["Fail_Filter_noempty"] = ratio_df[fail_filter].sum(axis=1)
+ratio_df["Pass/Fail_withempty"] = ratio_df["Pass_Filter"] / ratio_df["Fail_Filter"]
+ratio_df["Pass/Fail_0empty"] = ratio_df["Pass_Filter"] / ratio_df["Fail_Filter_noempty"]
+ratio_df["Percent_Empty"] = ratio_df["Empty"] / ratio_df["Sum"]
+ratio_df = ratio_df.drop(
+    cell_category_order + ["Sum", "Pass_Filter", "Fail_Filter", "Fail_Filter_noempty"],
+    1,
+)
+ratio_df = ratio_df.stack().to_frame().reset_index().rename(columns={0: "value"})
```

Comment:
lots of data manipulation going on here which we should try to get into a chain and `.assign()`

(optional btw, we can do later when I go in and refactor all the things)